File: common/src/main/java/com/jon/cot/common/service/Point.java
package com.jon.cot.common.service;

import com.jon.cot.common.cot.CursorOnTarget;
import com.jon.cot.common.utils.Constants;

public class Point {
    /* Both in radians */
    public double lat;
    public double lon;

    public Point(final double lat, final double lon) {
        this.lat = lat;
        this.lon = lon;
    }

    public Point add(Offset offset) {
        final double rOverR = offset.R / Constants.EARTH_RADIUS;
        final double theta = Math.toRadians(offset.theta);  // travel bearing in radians
        final double lat1 = this.lat;                       // start latitude in radians
        final double lon1 = this.lon;                       // start longitude in radians
        final double lat2 = Math.asin(Math.sin(lat1) * Math.cos(rOverR) + Math.cos(lat1) * Math.sin(rOverR) * Math.cos(theta));
        final double lon2 = lon1 + Math.atan2(Math.sin(theta) * Math.sin(rOverR) * Math.cos(lat1), Math.cos(rOverR) - Math.sin(lat1) * Math.sin(lat2));
        return new Point(lat2, lon2);
    }

    public static Point fromCot(CursorOnTarget cot) {
        return new Point(cot.lat * Constants.DEG_TO_RAD, cot.lon * Constants.DEG_TO_RAD);
    }

    public static class Offset {
        public double R;     /* travel distance in metres */
        public double theta; /* bearing in degrees */

        public Offset(double R, double theta) {
            this.R = R;
            this.theta = theta;
        }
    }
}

File: common/src/main/java/com/jon/common/service/Point.kt
package com.jon.common.service

import com.jon.common.cot.CursorOnTarget
import com.jon.common.utils.Constants
import kotlin.math.asin
import kotlin.math.atan2
import kotlin.math.cos
import kotlin.math.sin

data class Point(
        var lat: Double, // radians
        var lon: Double // radians
) {
    fun add(offset: Offset): Point {
        val rOverR = offset.R / Constants.EARTH_RADIUS_METRES
        val theta = Math.toRadians(offset.theta) // travel bearing in radians
        val lat1 = lat // start latitude in radians
        val lon1 = lon // start longitude in radians
        val lat2 = asin(sin(lat1) * cos(rOverR) + cos(lat1) * sin(rOverR) * cos(theta))
        val lon2 = lon1 + atan2(sin(theta) * sin(rOverR) * cos(lat1), cos(rOverR) - sin(lat1) * sin(lat2))
        return Point(lat2, lon2)
    }

    data class Offset(
            var R: Double,    /* travel distance in metres */
            var theta: Double /* bearing in degrees */
    )

    companion object {
        fun fromCot(cot: CursorOnTarget): Point {
            return Point(
                    cot.lat * Constants.DEG_TO_RAD,
                    cot.lon * Constants.DEG_TO_RAD
            )
        }
    }
}
